File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def full_house_rule(sorted_dice):
    if (sorted_dice[0] != sorted_dice[4]) and \
            ((sorted_dice[0] == sorted_dice[2] and sorted_dice[3] == sorted_dice[4])
             or (sorted_dice[0] == sorted_dice[1] and sorted_dice[2] == sorted_dice[4])):
        return sum(sorted_dice)
    else:
        return 0


def four_of_a_kind_rule(sorted_dice):
    if sorted_dice[0] == sorted_dice[3]:
        return sum(sorted_dice[0:4])
    elif sorted_dice[1] == sorted_dice[4]:
        return sum(sorted_dice[1:])
    else:
        return 0


scoring_rules = {
    YACHT: lambda sorted_dice: 50 if sorted_dice[0] == sorted_dice[4] else 0,
    ONES: lambda sorted_dice: sum([die for die in sorted_dice if die is 1]),
    TWOS: lambda sorted_dice: sum([die for die in sorted_dice if die is 2]),
    THREES: lambda sorted_dice: sum([die for die in sorted_dice if die is 3]),
    FOURS: lambda sorted_dice: sum([die for die in sorted_dice if die is 4]),
    FIVES: lambda sorted_dice: sum([die for die in sorted_dice if die is 5]),
    SIXES: lambda sorted_dice: sum([die for die in sorted_dice if die is 6]),
    FULL_HOUSE: full_house_rule,
    FOUR_OF_A_KIND: four_of_a_kind_rule,
    LITTLE_STRAIGHT: lambda sorted_dice: 30 if len(set(sorted_dice)) == len(sorted_dice)
                                               and sorted_dice[0] == 1 and sorted_dice[4] == 5 else 0,
    BIG_STRAIGHT: lambda sorted_dice: 30 if len(set(sorted_dice)) == len(sorted_dice)
                                               and sorted_dice[0] == 2 and sorted_dice[4] == 6 else 0,
    CHOICE: sum,
}


def score(dice, category):
    return scoring_rules[category](sorted(dice))

File: haskell/yacht/src/Yacht.hs
module Yacht (yacht, Category(..)) where

import Data.List (group, sort)

data Category = Ones
              | Twos
              | Threes
              | Fours
              | Fives
              | Sixes
              | FullHouse
              | FourOfAKind
              | LittleStraight
              | BigStraight
              | Choice
              | Yacht

yacht :: Category -> [Int] -> Int
yacht c dice = case c of
    Ones            -> numbers 1 dice
    Twos            -> numbers 2 dice
    Threes          -> numbers 3 dice
    Fours           -> numbers 4 dice
    Fives           -> numbers 5 dice
    Sixes           -> numbers 6 dice
    FullHouse       -> fullHouseScore dice
    FourOfAKind     -> fourOfAKindScore dice
    LittleStraight  -> if sort dice == [1, 2, 3, 4, 5] then 30 else 0
    BigStraight     -> if sort dice == [2, 3, 4, 5, 6] then 30 else 0
    Choice          -> sum dice
    Yacht           -> if length (group dice) == 1 then 50 else 0
    where numbers n = sum . filter (== n)

fullHouseScore :: [Int] -> Int
fullHouseScore dice = case group . sort $ dice of
    [[_,_,_],[_,_]] -> sum dice
    [[_,_],[_,_,_]] -> sum dice
    _               -> 0

fourOfAKindScore :: [Int] -> Int
fourOfAKindScore dice = case group . sort $ dice of
    [[_,_,_,_,x]]      -> sum dice - x
    [d@[_,_,_,_], [_]] -> sum d
    [[_], d@[_,_,_,_]] -> sum d
    _                  -> 0